File: export-templates/javascript/src/components/logger.js
const fs = require('fs');
const path = require('path');
const { mkdirpSync } = require('mkdirp');
const log4js = require('log4js');
const { AsyncLocalStorage } = require('async_hooks');
const logDir = path.join(process.cwd(), 'logs');
const logsFilePath = `${logDir}/app.log`;
const errorLogsFilePath = `${logDir}/err.log`;
const { LOG_LEVEL } = require('../constants');
const IS_LAMBDA = !!process.env.AWS_LAMBDA_FUNCTION_NAME;

if (!IS_LAMBDA) mkdirpSync(logDir);

// Create AsyncLocalStorage for request context
const asyncLocalStorage = new AsyncLocalStorage();

function setRequestContext(context) {
  asyncLocalStorage.enterWith(context);
}

// Custom JSON layout for structured logs
log4js.addLayout('structuredJson', function (config) {
  return function (logEvent) {
    const context = asyncLocalStorage.getStore() || {};
    const [first, ...rest] = logEvent.data || [];
    let message = '';
    let data = undefined;
    if (typeof first === 'string') {
      message = first;
      if (rest.length) {
        data = rest.length === 1 ? rest[0] : rest;
      }
    } else if (first !== undefined) {
      message = 'No message provided';
      data = [first, ...rest].length === 1 ? first : [first, ...rest];
    }
    // Remove requestId from data if present
    if (data && typeof data === 'object' && data.requestId) {
      data = { ...data };
      delete data.requestId;
    }
    const logObject = {
      timestamp: new Date(logEvent.startTime).toISOString(),
      level: logEvent.level.levelStr.toUpperCase(),
      category: logEvent.categoryName,
      message,
      requestId: context.requestId,
      data,
    };
    // Remove undefined fields
    Object.keys(logObject).forEach((key) => {
      if (logObject[key] === undefined) {
        delete logObject[key];
      }
    });
    return JSON.stringify(logObject) + (config.separator || '\n');
  };
});

const appenders = IS_LAMBDA
  ? {
      out: { type: 'stdout', layout: { type: 'structuredJson', separator: '\n' } },
      err: { type: 'stderr', layout: { type: 'structuredJson', separator: '\n' } },
    }
  : {
      out: { type: 'stdout', layout: { type: 'structuredJson', separator: '\n' } },
      err: { type: 'stderr', layout: { type: 'structuredJson', separator: '\n' } },
      file: {
        type: 'file',
        filename: logsFilePath,
        layout: { type: 'structuredJson', separator: ',' },
      },
      errorFile: {
        type: 'file',
        filename: errorLogsFilePath,
        layout: { type: 'structuredJson', separator: ',' },
      },
    };

const categories = IS_LAMBDA
  ? {
      default: { appenders: ['out', 'err'], level: 'info' },
      error: { appenders: ['err'], level: 'error' },
    }
  : {
      default: { appenders: ['out', 'err', 'file'], level: 'info' },
      error: { appenders: ['err', 'errorFile'], level: 'error' },
    };

log4js.configure({ appenders, categories });

const appLogger = log4js.getLogger();
const errorLogger = log4js.getLogger('error');

function log(message, ...args) {
  appLogger.info(message, ...args);
}

function logWarning(message, ...args) {
  appLogger.warn(message, ...args);
}

function logErr(message, ...args) {
  errorLogger.error(message, ...args);
}

function logDebug(message, ...args) {
  if (config.logging.LOG_LEVEL === 'debug') {
    appLogger.debug(message, ...args);
  }
}

function createAdhocLogFile(description, code, result, error = null) {
  const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
  const logFileName = path.join(logDir, `adhoc-${timestamp}${error ? '-error' : ''}.log`);
  const delimiter = '\n\n---\n\n';
  let prettyResult;
  try {
    const parsedResult = JSON.parse(result);
    prettyResult = JSON.stringify(parsedResult, null, 2);
  } catch (e) {
    prettyResult = result;
  }
  const logContent = [
    `Description: \n\n ${description}`,
    delimiter,
    `Code: \n\n ${code}`,
    delimiter,
    `Result: \n\n ${prettyResult}`,
    error ? `${delimiter}Error: ${error.stack}` : '',
  ].join('');
  fs.writeFileSync(logFileName, logContent);
}

module.exports = {
  log,
  logWarning,
  logErr,
  logDebug,
  createAdhocLogFile,
  setRequestContext,
};

File: examples/bank-report-demo-py/export/src/components/logger.py
import os
import json
import logging
from datetime import datetime
from logging.handlers import RotatingFileHandler
from constants import IS_LAMBDA

log_dir = os.path.join(os.getcwd(), 'logs')
logs_file_path = os.path.join(log_dir, 'app.log')
error_logs_file_path = os.path.join(log_dir, 'err.log')

if not IS_LAMBDA:
    os.makedirs(log_dir, exist_ok=True)

class JsonFormatter(logging.Formatter):
    def format(self, record):
        log_record = {
            'timestamp': self.formatTime(record, self.datefmt),
            'level': record.levelname,
            'message': record.getMessage(),
        }
        if record.exc_info:
            log_record['exception'] = self.formatException(record.exc_info)
        return json.dumps(log_record)

app_logger = logging.getLogger('app')
app_logger.setLevel(logging.INFO)
app_logger.propagate = False

error_logger = logging.getLogger('error')
error_logger.setLevel(logging.ERROR)
error_logger.propagate = False

if IS_LAMBDA:
    # Use stdout for Lambda logging
    stdout_handler = logging.StreamHandler()
    stdout_handler.setFormatter(JsonFormatter())
    app_logger.addHandler(stdout_handler)
    error_logger.addHandler(stdout_handler)
else:
    # Use file handlers for local environment
    app_file_handler = RotatingFileHandler(logs_file_path, maxBytes=10*1024*1024, backupCount=5)
    app_file_handler.setFormatter(JsonFormatter())
    app_logger.addHandler(app_file_handler)

    error_file_handler = RotatingFileHandler(error_logs_file_path, maxBytes=10*1024*1024, backupCount=5)
    error_file_handler.setFormatter(JsonFormatter())
    error_logger.addHandler(error_file_handler)

def log(command, event, body):
    app_logger.info(json.dumps({
        'command': command,
        'event': event,
        'body': body
    }))

def log_err(command, event, body, error):
    log_error = str(error) if error else None
    error_logger.error(json.dumps({
        'command': command,
        'event': event,
        'body': body,
        'error': log_error
    }))

def create_adhoc_log_file(description, code, result, error=None):
    timestamp = datetime.now().isoformat().replace(':', '-').replace('.', '-')
    log_file_name = os.path.join(log_dir, f"adhoc-{timestamp}{'-error' if error else ''}.log")
    
    delimiter = '\n\n---\n\n'

    if isinstance(result, dict):
        pretty_result = json.dumps(result, indent=2)
    else:
        try:
            # Try to parse result as JSON if it's a string
            pretty_result = json.dumps(json.loads(result), indent=2)
        except (json.JSONDecodeError, TypeError):
            # If parsing fails or result is not a string, use it as is
            pretty_result = str(result)

    log_content = delimiter.join([
        f"Description: \n\n {description}",
        f"Code: \n\n {code}",
        f"Result: \n\n {pretty_result}",
        f"Error: {error}" if error else ''
    ])

    with open(log_file_name, 'w') as f:
        f.write(log_content)